ir_config_parameter_multi_company/models/ir_config_parameter.py

```diff
@@ -10,39 +9,17 @@ class IrConfigMultiCompany(models.Model):
         ("key_uniq", "unique (key, company_id)", "Key must be unique per company.")
     ]
 
-    # this is needed because get_param is usually called with sudo
-    @api.model
-    def search(self, args, offset=0, limit=None, order=None, count=False):
-        if count:
-            return super().search(args, offset, limit, order, count)
-        key_present = False
-        for item in args:
-            if len(item) == 3 and item[0] == "key":
-                key_present = True
-        if not key_present:
-            return super().search(args, offset, limit, order, count)
-        else:
-            # avoid limit to fetch all values
-            records = super().search(
-                args=args, offset=offset, limit=None, order=order, count=count
-            )
-            if len(records) > 1:
-                # key must be unique per company
-                args = expression.AND(
-                    (args, [("company_id", "=", self.env.company.id)])
-                )
-                records = super().search(
-                    args=args, offset=offset, limit=None, order=order, count=count
-                )
-                if len(records) > 1:
-                    # call with limit
-                    records = super().search(
-                        args=args, offset=offset, limit=limit, order=order, count=count
-                    )
-        return records
-
     @api.model
     def _get_param(self, key):
-        # because self.env.company.id is not in cache key
-        self.clear_caches()
-        return super()._get_param(key)
+        company_id = self.env.company.id
+        if company_id:
+            self.flush(["key", "value", "company_id"])
+            self.env.cr.execute(
+                "SELECT value FROM ir_config_parameter WHERE key = %s "
+                "AND (company_id = %s OR company_id IS NULL)",
+                [key, company_id],
+            )
+            result = self.env.cr.fetchone()
```

Comment:
This does not guarantee that company value is used if both global and company specific value exists.

You can simply add ORDER BY company_id

NULL value will be at the end.

Reply:
Hi @yvaucher thanks, added.

I also added `def set_param` to avoid `Expected singleton` while writing a parameter already present for different companies